homeassistant/components/screenlogic/data.py

```diff
@@ -0,0 +1,313 @@
+"""Data constants for the ScreenLogic integration."""
+from abc import ABC, abstractmethod
+from collections.abc import Callable
+from dataclasses import dataclass
+import logging
+
+from screenlogicpy import ScreenLogicGateway
+from screenlogicpy.const.data import ATTR, DEVICE, VALUE
+from screenlogicpy.const.msg import CODE
+from screenlogicpy.device_const.system import EQUIPMENT_FLAG
+
+from homeassistant.backports.enum import StrEnum
+from homeassistant.const import EntityCategory
+
+from .const import SL_UNIT_TO_HA_UNIT, generate_unique_id
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class EntityParameter(StrEnum):
+    """Parameters for sensor mapping."""
+
+    ENABLED = "enabled"  # ScreenLogicRule
+    ENTITY_CATEGORY = "entity_category"  # EntityCategory
+    INCLUDED = "included"  # ScreenLogicRule
+    SET_VALUE = "set_value"  # tuple
+    SUBSCRIPTION_CODE = "sub_code"  # int
+    VALUE_MODIFICATION = "value_mod"  # Callable
+
+
+class PathPart(StrEnum):
+    """Placeholders for local data_path values."""
+
+    DEVICE = "!device"
+    KEY = "!key"
+    INDEX = "!index"
+    VALUE = "!sensor"
+
+
+class ScreenLogicRule(ABC):
+    """Base class for checking rules against a ScreenlogicGateway."""
+
+    def __init__(self, test: Callable[..., bool]) -> None:
+        """Initialize a ScreenLogic rule."""
+        self._test = test
+
+    @abstractmethod
+    def test(
+        self, gateway: ScreenLogicGateway, data_path: tuple[str | int, ...]
+    ) -> bool:
+        """Abstract method to check the rule."""
+
+
+class ScreenLogicDataRule(ScreenLogicRule):
+    """Represents a data rule."""
+
+    def __init__(
+        self, test: Callable[..., bool], test_path_template: tuple[str | int, ...]
+    ) -> None:
+        """Initialize a ScreenLogic data rule."""
+        self._test_path_template = test_path_template
+        super().__init__(test)
+
+    def test(
+        self, gateway: ScreenLogicGateway, data_path: tuple[str | int, ...]
+    ) -> bool:
+        """Check the rule against the gateway's data."""
+        test_path = realize_path_template(self._test_path_template, data_path)
+        return self._test(gateway.get_data(*test_path))
+
+
+class ScreenLogicEquipmentRule(ScreenLogicRule):
+    """Represents an equipment flag rule."""
+
+    def test(
+        self, gateway: ScreenLogicGateway, data_path: tuple[str | int, ...]
+    ) -> bool:
+        """Check the rule against the gateway's equipment flags."""
+        return self._test(gateway.equipment_flags)
+
+
+SupportedValueParameters = dict[
+    EntityParameter,
+    Callable | EntityCategory | int | ScreenLogicRule | tuple,
+]
+
+SupportedValueDescriptions = dict[str, SupportedValueParameters | dict]
+
+SupportedGroupDescriptions = dict[int | str, SupportedValueDescriptions]
+
+SupportedDeviceDescriptions = dict[str, SupportedGroupDescriptions]
+
```

Comment:
Are these used?

Reply:
Looks like its all to build `SupportedDeviceDescriptions` ...

Reply:
Yes. The goal was to make the type hint for `SUPPORTED_DATA` more readable as nesting `dict[]` became very long. Eventually I want to make `DEVICE`, `GROUP`, and `VALUE` enums, and replace `str` with them.